In [78]:
from pydataset import data

In [79]:
# data('mpg', show_doc=True) # view the documentation for the dataset
mpg = data('mpg') # load the dataset and store it in a variable


In [80]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [81]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)


pandas.core.frame.DataFrame

In [82]:
df

name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72

In [83]:
# 1.a.

df['passing_english'] = df.english >= 70
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72            False

In [84]:
# 1.b.

df.sort_values(by='passing_english', ascending=False) 

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False

In [85]:
# 1.c.

df.sort_values(by='name').sort_values(by='passing_english')

name  math  english  reading  passing_english
11     Alan    92       62       72            False
8    Albert    92       62       87            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True

In [86]:
# 1.d.

df.sort_values(by='english', ascending = False).sort_values(by='passing_english')

name  math  english  reading  passing_english
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False
10    Isaac    92       99       93             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
0     Sally    62       85       80             True
9   Richard    69       80       94             True
1      Jane    88       79       67             True
5      John    79       76       93             True
2     Suzie    94       74       95             True

In [87]:
# 1.e.

df['Average'] = (df.math + df.english + df.reading) / 3

In [88]:
df

name  math  english  reading  passing_english    Average
0     Sally    62       85       80             True  75.666667
1      Jane    88       79       67             True  78.000000
2     Suzie    94       74       95             True  87.666667
3     Billy    98       96       88             True  94.000000
4       Ada    77       92       98             True  89.000000
5      John    79       76       93             True  82.666667
6    Thomas    82       64       81            False  75.666667
7     Marie    93       63       90            False  82.000000
8    Albert    92       62       87            False  80.333333
9   Richard    69       80       94             True  81.000000
10    Isaac    92       99       93             True  94.666667
11     Alan    92       62       72            False  75.333333

In [89]:
from pydataset import data

In [90]:
# data('mpg', show_doc=True) # view the documentation for the dataset
mpg = data('mpg') # load the dataset and store it in a variable


In [91]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

How many rows and columns are there? 234 rows x 11 columns

In [92]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

In [93]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [94]:
mpg = mpg.rename(columns = {'cty': 'city'})

In [95]:
mpg.rename(columns = {'hwy': 'highway'})

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [96]:
mpg = mpg.rename(columns = {'hwy': 'highway'})

In [97]:
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [98]:
mpg['better_cpg'] = mpg.city - mpg.highway > 0

In [99]:
mpg.sort_values('better_cpg', ascending = True)

manufacturer                model  displ  year  cyl       trans drv  city  \
1           audi                   a4    1.8  1999    4    auto(l5)   f    18   
149       nissan               maxima    3.0  1999    6  manual(m5)   f    19   
150       nissan               maxima    3.5  2008    6    auto(av)   f    19   
151       nissan       pathfinder 4wd    3.3  1999    6    auto(l4)   4    14   
152       nissan       pathfinder 4wd    3.3  1999    6  manual(m5)   4    15   
..           ...                  ...    ...   ...  ...         ...  ..   ...   
84          ford      f150 pickup 4wd    4.2  1999    6    auto(l4)   4    14   
85          ford      f150 pickup 4wd    4.2  1999    6  manual(m5)   4    14   
86          ford      f150 pickup 4wd    4.6  1999    8  manual(m5)   4    13   
74         dodge  ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4    11   
234   volkswagen               passat    3.6  2008    6    auto(s6)   f    17   

     highway fl    class  better_cpg  
1         29  p  compact       False  
149       25  r  midsize       False  
150       25  p  midsize       False  
151       17  r      suv       False  
152       17  r      suv       False  
..       ... ..      ...         ...  
84        17  r   pickup       False  
85        17  r   pickup       False  
86        16  r   pickup       False  
74        15  r   pickup       False  
234       26  p  midsize       False  

[234 rows x 12 columns]

In [100]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   city          234 non-null    int64  
 8   highway       234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
 11  better_cpg    234 non-null    bool   
dtypes: bool(1), float64(1), int64(4), object(6)
memory usage: 22.2+ KB


In [101]:
mpg['mileage_difference'] = mpg.highway - mpg.city

In [102]:
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  better_cpg  mileage_difference  
1    compact       False                  11  
2    compact       False                   8  
3    compact       False                  11  
4    compact       False                   9  
5    compact       False                  10  
..       ...         ...                 ...  
230  midsize       False                   9  
231  midsize       False                   8  
232  midsize       False                  10  
233  midsize       False                   8  
234  midsize       False                   9  

[234 rows x 13 columns]

In [103]:
# 2. 

mpg.sort_values('mileage_difference', ascending = False)

# Honda Civic
# Volkswagon Beetle

manufacturer              model  displ  year  cyl       trans drv  city  \
107        honda              civic    1.8  2008    4    auto(l5)   f    24   
223   volkswagen         new beetle    1.9  1999    4    auto(l4)   f    29   
1           audi                 a4    1.8  1999    4    auto(l5)   f    18   
229   volkswagen             passat    1.8  1999    4    auto(l5)   f    18   
36     chevrolet             malibu    3.5  2008    6    auto(l4)   f    18   
..           ...                ...    ...   ...  ...         ...  ..   ...   
80          ford       explorer 4wd    4.0  1999    6    auto(l5)   4    14   
138      mercury    mountaineer 4wd    4.0  1999    6    auto(l5)   4    14   
177       toyota        4runner 4wd    3.4  1999    6  manual(m5)   4    15   
152       nissan     pathfinder 4wd    3.3  1999    6  manual(m5)   4    15   
204       toyota  toyota tacoma 4wd    3.4  1999    6  manual(m5)   4    15   

     highway fl       class  better_cpg  mileage_difference  
107       36  c  subcompact       False                  12  
223       41  d  subcompact       False                  12  
1         29  p     compact       False                  11  
229       29  p     midsize       False                  11  
36        29  r     midsize       False                  11  
..       ... ..         ...         ...                 ...  
80        17  r         suv       False                   3  
138       17  r         suv       False                   3  
177       17  r         suv       False                   2  
152       17  r         suv       False                   2  
204       17  r      pickup       False                   2  

[234 rows x 13 columns]

In [104]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   manufacturer        234 non-null    object 
 1   model               234 non-null    object 
 2   displ               234 non-null    float64
 3   year                234 non-null    int64  
 4   cyl                 234 non-null    int64  
 5   trans               234 non-null    object 
 6   drv                 234 non-null    object 
 7   city                234 non-null    int64  
 8   highway             234 non-null    int64  
 9   fl                  234 non-null    object 
 10  class               234 non-null    object 
 11  better_cpg          234 non-null    bool   
 12  mileage_difference  234 non-null    int64  
dtypes: bool(1), float64(1), int64(5), object(6)
memory usage: 24.0+ KB


In [105]:
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  better_cpg  mileage_difference  
1    compact       False                  11  
2    compact       False                   8  
3    compact       False                  11  
4    compact       False                   9  
5    compact       False                  10  
..       ...         ...                 ...  
230  midsize       False                   9  
231  midsize       False                   8  
232  midsize       False                  10  
233  midsize       False                   8  
234  midsize       False                   9  

[234 rows x 13 columns]

In [106]:
mpg.sort_values('highway', ascending = False)

# Volkwagon Jetta

manufacturer                model  displ  year  cyl       trans drv  city  \
222   volkswagen           new beetle    1.9  1999    4  manual(m5)   f    35   
213   volkswagen                jetta    1.9  1999    4  manual(m5)   f    33   
223   volkswagen           new beetle    1.9  1999    4    auto(l4)   f    29   
197       toyota              corolla    1.8  2008    4  manual(m5)   f    28   
106        honda                civic    1.8  2008    4    auto(l5)   f    25   
..           ...                  ...    ...   ...  ...         ...  ..   ...   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4     9   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   

     highway fl       class  better_cpg  mileage_difference  
222       44  d  subcompact       False                   9  
213       44  d     compact       False                  11  
223       41  d  subcompact       False                  12  
197       37  r     compact       False                   9  
106       36  r  subcompact       False                  11  
..       ... ..         ...         ...                 ...  
66        12  e      pickup       False                   3  
60        12  e         suv       False                   3  
127       12  e         suv       False                   3  
70        12  e      pickup       False                   3  
55        12  e      pickup       False                   3  

[234 rows x 13 columns]

In [107]:
mpg['average_mileage'] = mpg.city + mpg.highway / 2



In [108]:
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  better_cpg  mileage_difference  average_mileage  
1    compact       False                  11             32.5  
2    compact       False                   8             35.5  
3    compact       False                  11             35.5  
4    compact       False                   9             36.0  
5    compact       False                  10             29.0  
..       ...         ...                 ...              ...  
230  midsize       False                   9             33.0  
231  midsize       False                   8             35.5  
232  midsize       False                  10             29.0  
233  midsize       False                   8             31.0  
234  midsize       False                   9             30.0  

[234 rows x 14 columns]

In [111]:
mpg.sort_values('average_mileage', ascending = False)

manufacturer                model  displ  year  cyl       trans drv  city  \
222   volkswagen           new beetle    1.9  1999    4  manual(m5)   f    35   
213   volkswagen                jetta    1.9  1999    4  manual(m5)   f    33   
223   volkswagen           new beetle    1.9  1999    4    auto(l4)   f    29   
197       toyota              corolla    1.8  2008    4  manual(m5)   f    28   
100        honda                civic    1.6  1999    4  manual(m5)   f    28   
..           ...                  ...    ...   ...  ...         ...  ..   ...   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4     9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   

     highway fl       class  better_cpg  mileage_difference  average_mileage  
222       44  d  subcompact       False                   9             57.0  
213       44  d     compact       False                  11             55.0  
223       41  d  subcompact       False                  12             49.5  
197       37  r     compact       False                   9             46.5  
100       33  r  subcompact       False                   5             44.5  
..       ... ..         ...         ...                 ...              ...  
66        12  e      pickup       False                   3             15.0  
55        12  e      pickup       False                   3             15.0  
70        12  e      pickup       False                   3             15.0  
127       12  e         suv       False                   3             15.0  
60        12  e         suv       False                   3             15.0  

[234 rows x 14 columns]